lib/watchers/fs-watcher.js

```diff
@@ -0,0 +1,439 @@
+            }
+        }
+
+        vasync.forEachParallel({
+            inputs: vms,
+            func: function _statXmlFile(vm, cb) {
+                fs.stat('/etc/zones/' + vm + '.xml', function _onStat(e, st) {
+                    if (e) {
+                        if (e.code === 'ENOENT') {
+                            // If file disappeared, that's fine.
+                            cb();
+                            return;
+                        }
+                        cb(e);
+                        return;
+                    }
+
+                    self.getConfigTimestamp(vm,
+                        function _getConfigTimestampCb(error, latestCfg) {
+                            if (error) {
+                                cb(error);
+                                return;
+                            }
+                            // Thanks Javascript... NaN > X, where X is a
+                            // number. So we '|| 0' here to work around it.
+                            results[vm] = Math.max(st.mtime.getTime() || 0,
+                                latestCfg || 0);
+                            cb();
+                        }
+                    );
+                });
+            }
+        }, function _afterStats(e) {
+            var doneLookup = (new Date()).getTime();
+
+            self.log.trace({
+                action: 'getCurrentZones',
+                elapsed: (doneLookup - startLookup),
+                err: e,
+                vmCount: (vms ? Object.keys(vms).length : 0)
+            }, 'completed lookup of latest timestamps');
+
+            callback(e, results);
+        });
+    });
+};
+
+FsWatcher.prototype.getConfigTimestamp = // eslint-disable-line
+function getConfigTimestamp(vm, callback) {
+    var self = this;
+    var timestamps = [];
+
+    vasync.forEachParallel({
+        inputs: [
+            '/zones/' + vm + '/config/metadata.json',
+            '/zones/' + vm + '/config/routes.json',
+            '/zones/' + vm + '/config/tags.json'
+        ],
+        func: function _statConfigFile(filename, cb) {
+            fs.stat(filename, function _onStat(e, st) {
+                if (!e) {
+                    timestamps.push(st.mtime.getTime());
+                } else if (e.code !== 'ENOENT') {
+                    // If stat has an error we'll just log it since there's not
+                    // much else for us to do about it.
+                    self.log.warn(e, 'fs.stat error');
+                }
+                cb();
+            });
+        }
+    }, function _afterConfigFileStats(err) {
+        var latest = timestamps.sort().pop();
+
+        callback(err, latest);
+    });
+};
+
+/*
+ * For config we only watch for change events because when a VM is created we'll
+ * always see the .xml file but we may or may not see config. Since xml is
+ * always there we'll just rely on that.
+ *
+ */
+FsWatcher.prototype.watchConfig = function watchConfig(vmUuid, callback) {
```

Comment:
`FSWatcher.prototype.watchConfig` doesn't seem to need to take a `callback`, as it just schedules async I/O, but doesn't wait for them to complete. I'm wondering if it would be clearer to write it as a function not taking a callback.